### DISCUSSION TITLE
add etcd api v3 support

### DISCUSSION THREAD

**Discussion by jcarnu** [cybertec-postgresql/pgwatch2#782] [Ideas]
Hello,

I wrote a first try on etcd v3 api support in pgwatch2. I was wondering if there was a way to test it before submitting a PR and make some debugging with all the dockers things in the source code (sorry to bother you with that, maybe it worth asking you before going on writing my own test bench) ?

After a first quick reading  it seems there's none.

Would you like me to submit a PR at first place before doing any testing actions (it compiles fine) so you could also make remarks on code and design ?

Roughly it's a simple hack of `EtcdGetClusterMembers()` and a simple duplication (bad) of `extractEtcdScopeMembers()`

Thanks.

**Comment by pashagolub** [comment]
Hello!

Thanks a lot for your interest! pgwatch2 is now in maintenance mode, accepting only bug fixes. Check out the WIP [pgwatch3](https://github.com/cybertec-postgresql/pgwatch3/) repository for new features and enhancements!

We already implemented etcd v3 support in pgwatch3

**Comment by jcarnu** [comment]
Hello @pashagolub ,

Thanks for your reply. I understand you're in full dev for pgwatch3. There's still no release for it (and that's quite normal). My point is that etcdv2 is tagged as deprecated and will be soon. 

Having pgwatch2 with no support of etcdv3  implies sticking to etcdv2 for all the patroni clusters. We so have the choice to have a nearly deprecated api risk and, OTOH being blind with no pgwatch2 monitoring which is also a (Ops) risk.

It sounds to me that this might be a good alternative to consider bringing etcdv3 to pgwatch2 even if it is in maintainance mode for that reason.... till pgwatch3 has a stable release. Is there any way or any thing I could do to get an exception on that ?

Don't get me wrong, this is only my point of view, and I know that maintaining/accepting features on a project implies a lot of work for you and the team and I don't want to insist on that point.

**Reply by pashagolub** [reply]
Of course you can maintain your own fork of pgwatch2. That is the power of open source.

The problem is you cannot switch to etcdv3 without supporting v2 for back compatibility. Otherwise you will blow up all current deployments.

Having both of them in one binary leads to a messy and dirty code. etcd v3 was one of the main reasons we started to work on a new major release.

Everything new goes to pgwatch3, e.g. Grafana 9.x, etcd v3, Postgres 17+, etc.

**Reply by jcarnu** [reply]
I agree with you, but I won't "maintain" a fork of the "official" version. This was not ma intention :) The proposed patch supported both etcv2 and etcv3 and you're right the code was quite well messy (to my standard also). 

Agreed on pgwatch3 targets.
